core/src/main/java/org/apache/gravitino/authorization/PermissionManager.java

```diff
@@ -373,11 +377,253 @@ User revokeRolesFromUser(String metalake, List<String> roles, String user) {
+                                              roleEntity,
+                                              RoleChange.addSecurableObject(role, securableObject));
+                                        }));
+
+                            return securableObject;
+                          } else {
+                            // Removed duplicated privileges by set
+                            Set<Privilege> updatePrivileges = Sets.newHashSet();
+                            updatePrivileges.addAll(oldObject.privileges());
+                            // If old object contains all the privileges to grant, the object don't
+                            // need to change.
+                            if (updatePrivileges.containsAll(privileges)) {
+                              return oldObject;
+                            } else {
+                              updatePrivileges.addAll(privileges);
+                              SecurableObject newSecurableObject =
+                                  SecurableObjects.parse(
+                                      oldObject.fullName(),
+                                      oldObject.type(),
+                                      Lists.newArrayList(updatePrivileges));
+
+                              authorizationPluginCallbackWrapper.setCallBack(
+                                  () ->
+                                      AuthorizationUtils.callAuthorizationPluginForMetadataObject(
+                                          metalake,
+                                          object,
+                                          authorizationPlugin -> {
+                                            authorizationPlugin.onRoleUpdated(
+                                                roleEntity,
+                                                RoleChange.updateSecurableObject(
+                                                    role, oldObject, newSecurableObject));
+                                          }));
+
+                              return newSecurableObject;
+                            }
+                          }
+                        });
+
+                AuditInfo auditInfo =
+                    AuditInfo.builder()
+                        .withCreator(roleEntity.auditInfo().creator())
+                        .withCreateTime(roleEntity.auditInfo().createTime())
+                        .withLastModifier(PrincipalUtils.getCurrentPrincipal().getName())
+                        .withLastModifiedTime(Instant.now())
+                        .build();
+
+                return RoleEntity.builder()
+                    .withId(roleEntity.id())
+                    .withName(roleEntity.name())
+                    .withNamespace(roleEntity.namespace())
+                    .withProperties(roleEntity.properties())
+                    .withAuditInfo(auditInfo)
+                    .withSecurableObjects(updateSecurableObjects)
+                    .build();
+              });
```

Comment:
I think the method is still too big, you need to split it into some methods.

Reply:
Already splitted.